logd/store/badger/history.go

```diff
@@ -71,18 +112,56 @@ func (log *Log) GetEvents(ctx context.Context, after string, limit int) ([]api.E
 					return err
 				}
 
-				events = append(events, evt)
+				events[curIndex] = evt
+				curIndex += indexDelta
+				eventsProcessed++
 				return nil
 			}); err != nil {
 				return err
 			}
-			eventsProcessed++
 		}
 		return nil
 	}); err != nil {
 		return nil, fmt.Errorf("scanning index: %w", err)
 	}
-	return events, nil
+
+	if direction == store.DirectionForward {
+		events = events[0:curIndex]
+	} else {
+		start := curIndex - indexDelta
+		events = events[start:]
+	}
+
+	eventsWithCursors := make([]store.EventWithCursors, len(events))
+	var err error
+	for i, event := range events {
+		if eventsWithCursors[i], err = getEventWithCursors(event); err != nil {
+			return nil, fmt.Errorf("creating cursors for event %q: %w", event.ID, err)
+		}
+	}
+
+	return eventsWithCursors, nil
+}
+
+func getEventWithCursors(event api.Event) (store.EventWithCursors, error) {
+	eventWithCursors := store.EventWithCursors{
+		Event: event,
+	}
+	idBytes, err := decodeID(event.ID)
+	if err != nil {
+		return store.EventWithCursors{}, err
+	}
+
+	// Copy the ID since increment/decrement operations mutate.
+	prevCursorID := make([]byte, len(idBytes))
+	copy(prevCursorID, idBytes)
+
+	// We don't care about the error if this is already a (zero-valued) NilCursor
+	_ = binaryutil.DecrementBytes(prevCursorID)
```

Comment:
suuuppper confusing that DrecrementBytes returns an error and mutates its argument, but IncrementBytes returns a new slice?

Reply:
`IncrementBytes` also mutates argument, but like `append`, it may need to reallocate. It cannot error unless it has to allocate, and `malloc` fails. `DecrementBytes`, however, has an underflow case.